modules/AlphaWalletFoundation/AlphaWalletFoundation/TokenScriptClient/Models/AssetDefinitionStore.swift

```diff
@@ -222,8 +222,8 @@ public class AssetDefinitionStore: NSObject {
 
         cancelable[request.hashValue] = network
             .fetchXml(request: request)
-            .sink(receiveCompletion: { [cancelable] _ in
-                cancelable[request.hashValue] = .none
+            .sink(receiveCompletion: { [weak self] _ in
+                self?.cancelable[request.hashValue] = .none
```

Comment:
Torn over situations like this. This means access to `cancelable` is on the thread that `.sink(receiveCompletion:)`runs on, which is dependent on what `.fetchXml(request:)` runs on (which is `.main` since there's a `receive(on:)` in the implementation) , unless we had a specific `receive(on:)` before it. This would probably run decently for us as we migrate most things to run on main, and then it'll just take 1 or a few calls that runs on non-main thread to cause bugs to appear again mysteriously.

Reply:
I don't know of a fullproof way, but I think what we agreed earlier to do — always "return" to the main thread is the best way. Kind of like what `fetchXml(request:)` does, so if a function has to branch off the non-main, it must always "return" back on main.

Nothing to change here. Just sharing some thoughts.